### PR TITLE
Publishing bam from sentieon-bwamem

### PR DESCRIPTION
**Pull request by asp8200**
The following test-cmd didn't publish any bam-files (just bai-files) from SENTIEON_BWAMEM:

```
nextflow run main.nf -profile test_cache,docker --aligner sentieon-bwamem --save_mapped --save_output_as_bam --outdir results
```
Now it does :-)  I tried to improve on some related stuff.  Added some pytests.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/modules/aligner.config [hunk 1]
```diff
@@ -44,8 +44,41 @@ process {
...
+                    (meta.size * meta.num_lanes == 1) ? "mapped/${meta.id}/${it}" : null
+                }
+            ]
+        }
+
+
+        withName: "SENTIEON_BWAMEM" {
+            // Markduplicates Spark NEEDS name-sorted reads or runtime goes through the roof.
+            // However, currently SENTIEON_BWAMEM only supports coordinate sorting the reads.
+            ext.prefix       = { params.split_fastq > 1 ? "${meta.id}".concat('.').concat(reads.get(0).name.tokenize('.')[0]) : "${meta.id}.sorted" }
```
Comment (asp8200) [inline conf/modules/aligner.config:66]: @maxulysse : The only difference between the config-snippet for `BWAMEM.*_MEM|DRAGMAP_ALIGN` and this one for `SENTIEON_DEDUP` is that the former uses `args2`. However, the module `SENTIEON_DEDUP` doesn't use `args2`.

As the whole logic here is a bit tricky, I decided to make two separate config-snippets instead one combined.

Comment (maxulysse) [reply conf/modules/aligner.config:66]: That sounds like a good idea

File: conf/modules/aligner.config [hunk 2]
```diff
@@ -44,8 +44,41 @@ process {
...
-                saveAs: { (params.save_output_as_bam && (params.save_mapped || params.skip_tools && params.skip_tools.split(',').contains('markduplicates'))) && (meta.size * meta.num_lanes == 1) ? "mapped/${meta.id}/${it}" : null }
+                // Only save if save_output_as_bam AND
+                // (save_mapped OR no_markduplicates OR sentieon_dedup) AND
+                // only a single BAM file per sample
+                saveAs: {
+                    params.save_output_as_bam &&
+                    (
+                        params.save_mapped ||
+                        (params.skip_tools && params.skip_tools.split(',').contains('markduplicates')) ||
+                        !(params.tools && params.tools.split(',').contains('sentieon_dedup'))
```
Comment (maxulysse) [inline conf/modules/aligner.config:55]: Would one use sentieon to map and not to dedup?

File: conf/modules/aligner.config [hunk 3]
```diff
@@ -44,8 +44,41 @@ process {
...
+                    ) &&
+                    (meta.size * meta.num_lanes == 1) ? "mapped/${meta.id}/${it}" : null }
```
Comment (maxulysse) [inline conf/modules/aligner.config:81]: I don't like having the ternary syntax on multiple lines like that, but for such a long condition, we might as well convert to a regular if

Comment (asp8200) [reply conf/modules/aligner.config:81]: How about this?

```
                saveAs: {
                    if (params.save_output_as_bam &&
                        (
                            params.save_mapped ||
                            (params.skip_tools && params.skip_tools.split(',').contains('markduplicates')) ||
                            !(params.tools && params.tools.split(',').contains('sentieon_dedup'))
                        ) && (meta.size * meta.num_lanes == 1)
                    ) { "mapped/${meta.id}/${it}" }
                    else { null }
                }
```

Comment (maxulysse) [reply conf/modules/aligner.config:81]: is it working?

**Review by maxulysse — APPROVED** [review]

**Pull request merged**